### PR TITLE
Node connection metrics

### PR DESCRIPTION
**Pull request by ShanniLi**
r @Raynos @kriskowal @mmihic 
cc @jcorbin 

last big piece of metrics. I will add the missing small pieces next, e.g., bytes sent/received

### CODE REVIEW

File: node/lib/statsd.js [hunk 1]
```diff
@@ -127,6 +127,39 @@ function getKey(common, stat) {
...
+
+        case 'connections.errors':
+            return prefix + '.' +
+                clean(stat.tags['peer-host-port'], 'peer-host-port') + '.' +
+                clean(stat.tags.type, 'type');
+
+        case 'connections.closed':
+            return prefix + '.' +
+                clean(stat.tags['peer-host-port'], 'peer-host-port') + '.' +
+                clean(stat.tags.reason, 'reason');
```
Comment (kriskowal) [inline node/lib/statsd.js]: I sense a pattern emerging that approximately `sprintf(patterns[stat.name], stat.tags)` might solve.

Comment (kriskowal) [inline node/lib/statsd.js]: (but don’t. might be costly)

**Comment by kriskowal** [comment]
:ok:

File: node/connection_base.js [hunk 2]
```diff
@@ -71,6 +73,31 @@ function TChannelConnectionBase(channel, direction, remoteAddr) {
 
     self.tracer = self.channel.tracer;
     self.ops.startTimeoutTimer();
+
+    if (direction === 'out') {
+        self.channel.connectionsInitiatedStat.increment(1, {
```
Comment (Raynos) [inline node/connection_base.js]: we do not want to report self connections.

Comment (Raynos) [inline node/connection_base.js]: just move all the code out of connection_base

File: node/connection_base.js [hunk 3]
```diff
@@ -71,6 +73,31 @@ function TChannelConnectionBase(channel, direction, remoteAddr) {
...
+    }
+
+    if (self.reportInterval > 0) {
+        self.reportTimer = self.timers.setTimeout(
+            onReport, self.reportInterval
+        );
+    }
+
+    function onReport() {
+        self.channel.connectionsActiveStat.update(1, {
```
Comment (Raynos) [inline node/connection_base.js]: That's not how gauges work.

You need to `update(totalNum)`.

The only person that has access to that is the rootChannel. It tracks all the incoming connections in serverConnections and it has access to all the peers so in theory can track outgoing connections.

File: node/connection_base.js [hunk 4]
```diff
@@ -104,6 +131,39 @@ TChannelConnectionBase.prototype.resetAll = function resetAll(err) {
...
     }
 
+    if (!self.remoteName) {
+        if (self.direction === 'out') {
+            self.channel.connectionsConnectErrorsStat.increment(1, {
+                'host-port': self.channel.hostPort || '0.0.0.0:0',
+                'peer-host-port': self.remoteAddr
+            });
+        } else {
+            self.channel.connectionsAcceptedErrorsStat.increment(1, {
```
Comment (Raynos) [inline node/connection_base.js]: we know the peer; let's add it and just modify the metrics document or ask clarification from @mmihic

File: node/connection_base.js [hunk 5]
```diff
@@ -104,6 +131,39 @@ TChannelConnectionBase.prototype.resetAll = function resetAll(err) {
...
+                'host-port': self.channel.hostPort
+            });
+        }
+    } else {
+        if (err.type !== 'tchannel.socket-local-closed' &&
+            err.type !== 'tchannel.socket-closed') {
```
Comment (Raynos) [inline node/connection_base.js]: socket-closed is also an error. The only one's that are not errors are local closes.

File: node/test/connection-stats.js [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+'use strict';
+
+var os = require('os');
+var TimeMock = require('time-mock');
+
+var allocCluster = require('./lib/alloc-cluster.js');
+var timers = TimeMock(Date.now());
```
Comment (Raynos) [inline node/test/connection-stats.js]: no; please create a timer per tests.

File: node/test/connection-stats.js [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+        app: 'server',
+        host: os.hostname()
+    };
+    client.on('stat', function onStat(stat) {
+        stats.push(stat);
+    });
+
+    var subClient = client.makeSubChannel({
+        serviceName: 'server',
+        peers: ['localhost:4040']
```
Comment (Raynos) [inline node/test/connection-stats.js]: leave a comment about there should be nothing running on this port.

File: node/test/connection-stats.js [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+                value: 1,
+                tags:
+                {
+                    'host-port': clientHost,
+                    'peer-host-port': serverHost,
+                    app: 'server',
+                    host: os.hostname()
+               }
+            }, {
+                name: 'connections.active',
```
Comment (Raynos) [inline node/test/connection-stats.js]: write a test where you open 3 connections and gauge is set to 3 :)

**Comment by Raynos** [comment]
Reviewed.

 - We have to fix the gauge. 
 - We should not report self connections

File: node/test/connection-stats.js (hunk 6 above)
Comment (ShanniLi) [inline node/test/connection-stats.js]: there is only one test uses that timer.

File: node/connection_base.js (hunk 3 above)
Comment (ShanniLi) [inline node/connection_base.js]: it is "The number of currently active outbound connections from the process to the peer.", so it is per peer bases. I can move the connection to peer level. It cannot be at the channel level by definition.

File: docs/metrics.md [hunk 9]
```diff
@@ -218,8 +218,7 @@ The total number of connections from the target peer accepted by this process
 over its lifetime. 
 
 ##### connections.accept-errors
-The total number of accept errors that have occurred on this process.  Will not
-include a peer tag.
+The total number of accept errors that have occurred on this process.
```
Comment (Raynos) [inline docs/metrics.md]: @mmihic we are including the peer tag in this one because we have access to it.

File: node/connection.js [hunk 10]
```diff
@@ -378,4 +391,111 @@ TChannelConnection.prototype.buildOutResponse = function buildOutResponse(req, o
     return self.handler.buildOutResponse(req, opts);
 };
 
+// this connection is completely broken, and is going away
+// In addition to erroring out all of the pending work, we reset the state
+// in case anybody stumbles across this object in a core dump.
+TChannelConnectionBase.prototype.resetAll = function resetAll(err) {
```
Comment (Raynos) [inline node/connection.js]: XD dont mutate the other prototype bro.

File: node/connection_base.js [hunk 11]
```diff
@@ -75,91 +75,9 @@ function TChannelConnectionBase(channel, direction, remoteAddr) {
 inherits(TChannelConnectionBase, EventEmitter);
 
 TChannelConnectionBase.prototype.close = function close(callback) {
-    var self = this;
-
-    self.resetAll(errors.LocalSocketCloseError());
```
Comment (Raynos) [inline node/connection_base.js]: you can't change this; Move this `close()` into `TChannelConnection`

Comment (Raynos) [inline node/connection_base.js]: Then just remove this method.

Comment (ShanniLi) [inline node/connection_base.js]: cannot do that, also used by self-connection ...

**Comment by Raynos** [comment]
lgtm

**Pull request merged**